File: java/AlgorithmAndDataStructure/src/test/java/algorithms/NonModifyingSequenceTest.java
package algorithms;

import functionObjects.EqualX;
import functionObjects.AppendToList;
import functionObjects.Equals;
import junit.framework.Test;
import junit.framework.TestCase;
import junit.framework.TestSuite;

import java.util.AbstractMap;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Unit test for simple NonModifyingSequence.
 */
public class NonModifyingSequenceTest
    extends TestCase
{
	private ArrayList<Integer> a;
	private EqualX equalThree;
	private Equals equals;

    /**
     * Create the test case
     *
     * @param testName name of the test case
     */
    public NonModifyingSequenceTest(String testName)
    {
        super( testName );
    }

    /**
     * @return the suite of tests being tested
     */
    public static Test suite()
    {
        return new TestSuite( NonModifyingSequenceTest.class );
    }

	public void setUp() {
		a = new ArrayList<Integer>();
		equalThree = new EqualX(3);
		equals = new Equals();
	}

    /**
     * Rigours Test :-)
     */
    public void testApp()
    {
        assertTrue( true );
    }

	public void testAnyOf() throws Exception {
		a.add(1);
		a.add(2);
		a.add(3);
		assertTrue(NonModifyingSequence.anyOf(a, equalThree));
		a.remove(Integer.valueOf(3));
		assertFalse(NonModifyingSequence.anyOf(a, equalThree));
	}

	public void testAllOf() throws Exception {
		a.add(3);
		a.add(3);
		assertTrue(NonModifyingSequence.allOf(a, equalThree));
		a.set(1, 2);
		assertFalse(NonModifyingSequence.allOf(a, equalThree));
	}

	public void testNoneOf() throws Exception {
		a.add(2);
		assertTrue(NonModifyingSequence.noneOf(a, equalThree));
		a.add(3);
		assertFalse(NonModifyingSequence.noneOf(a, equalThree));
	}

	public void testForEach() throws Exception {
		ArrayList<Integer> ary = new ArrayList<Integer>();
		ary.add(1);
		ary.add(2);
		AppendToList aToL = new AppendToList();
		NonModifyingSequence.forEach(ary, aToL);
		assertEquals(ary, aToL.getAry());
	}

	public void testCount_if() throws Exception {
		a.add(2);
		a.add(3);
		assertEquals(1, NonModifyingSequence.count_if(a, equalThree));
	}

	public void testCount() throws Exception {
		a.add(2);
		a.add(3);
		assertEquals(1, NonModifyingSequence.count(a, 3));
	}

	public void testMismatch() throws Exception {
		ArrayList<Integer> ary = new ArrayList<Integer>();
		ary.add(1);
		ary.add(2);
		a.add(2);
		a.add(3);
		AbstractMap.SimpleImmutableEntry entry = NonModifyingSequence.mismatch(a, ary, equals);
		assertEquals(3, ((Iterator)(entry.getKey())).next());
		assertEquals(2, ((Iterator)(entry.getValue())).next());
	}

	public void testEqual() throws Exception {
		ArrayList<Integer> ary = new ArrayList<Integer>();
		ary.add(1);
		a.add(1);
		assertTrue(NonModifyingSequence.equal(ary, a, equals));
		a.remove(Integer.valueOf(1));
		assertFalse(NonModifyingSequence.equal(ary, a, equals));
	}

	public void testFind() throws Exception {
		a.add(1);
		assertEquals(Integer.valueOf(1), NonModifyingSequence.find(a, 1));
	}

	public void testFindIf() throws Exception {
		a.add(3);
		assertEquals(Integer.valueOf(3), NonModifyingSequence.findIf(a, equalThree));
	}

	public void testFindIfNot() throws Exception {
		a.add(3);
		a.add(1);
		assertEquals(Integer.valueOf(1), NonModifyingSequence.findIfNot(a, equalThree));
	}

	public void testFindFirstOf() throws Exception {
		a.add(1);
		ArrayList<Integer> ary = new ArrayList<Integer>();
		ary.add(0);
		ary.add(1);
		assertFalse(NonModifyingSequence.findFirstOf(a, ary, equals).hasNext());
	}

	public void testFindEnd() throws Exception {
		a.add(1);
		a.add(1);
		ArrayList<Integer> ary = new ArrayList<Integer>();
		ary.add(1);
		assertEquals(1, NonModifyingSequence.findEnd(a, ary, equals));
		ary.add(1);
		assertEquals(0, NonModifyingSequence.findEnd(a, ary, equals));
	}

	public void testAdjacentFind() throws Exception {
		a.add(1);
		a.add(1);
		assertFalse(NonModifyingSequence.adjacent_find(a, equals).hasNext());
	}

	public void testSearch() throws Exception {
		a.add(1);
		a.add(1);
		ArrayList<Integer> ary = new ArrayList<Integer>();
		ary.add(1);
		assertEquals(0, NonModifyingSequence.search(a, ary, equals));
	}

	public void testSearchN() throws Exception {
		a.add(1);
		a.add(1);
		assertEquals(1, NonModifyingSequence.searchN(a, 2, Integer.valueOf(1), equals));
		assertEquals(2, NonModifyingSequence.searchN(a, 3, Integer.valueOf(1), equals));
	}
}

File: php/NonModifyingSequenceTest.php
<?php
namespace Algorithm;

class NonModifyingSequenceTest extends \PHPUnit_Framework_TestCase
{
    private $ary;

    /*
     * @before
     */
    public function setUp() {
        $this->ary = [1, 2, 3, 4, 6];
    }

    public function testAllOf()
    {
        $this->assertTrue(allOf($this->ary, function($x) {
            return $x > 0;
        }));
        $this->assertFalse(allOf($this->ary, function($x) {
            return $x < 3;
        }));
    }

    public function testAnyOf() {
        $this->assertTrue(anyOf($this->ary, function($x) {
            return $x < 2;
        }));
        $this->assertFalse(anyOf($this->ary, function($x) {
            return $x < 1;
        }));
    }

    public function testNoneOf()
    {
        $this->assertTrue(noneOf($this->ary, function($x) {
            return $x < 1;
        }));
        $this->assertFalse(noneOf($this->ary, function($x) {
            return $x < 6;
        }));
    }

    public function testForEach()
    {
        $func = function($x) {
            return $x + 1;
        };
        $newAry = array_map($func, $this->ary);
        for_each($this->ary, $func);
        $this->assertEquals($newAry, $this->ary);
    }

    public function testCount()
    {
        $this->assertEquals(1, count($this->ary, 2));
        $this->assertEquals(0, count($this->ary, 5));
    }

    public function testCountIf()
    {
        $this->assertEquals(1, countIf($this->ary, function($x) {
            return $x === 2;
        }));
    }

    public function testMismatch()
    {
        $ary = range(1, 5);
        $mismatchIdx = mismatch($this->ary, $ary);
        $this->assertEquals(4, $mismatchIdx);
        $newAry = $ary;
        $this->assertNull(mismatch($ary, $newAry));
    }

    public function testEqual()
    {
        $func = function($x) {
            return $x + 1;
        };
        $newAry = array_map($func, $this->ary);
        $cp = function($x, $y) {
            return $x === $y + 1;
        };
        $this->assertFalse(equal($this->ary, $newAry, $cp));
        $this->assertTrue(equal($newAry, $this->ary, $cp));
    }

    public function testFind()
    {
        $idx = find($this->ary, 3);
        $this->assertEquals(2, $idx);
    }

    public function testFindIf()
    {
        $this->assertEquals(2, findIf($this->ary, function($x) {
            return $x > 2;
        }));
        $this->assertNull(findIf($this->ary, function($x) {
            return $x > 6;
        }));
    }

    public function testFindIfNot()
    {
        $this->assertEquals(2, findIfNot($this->ary, function($x) {
            return $x < 3;
        }));
    }

    public function testFindEnd()
    {
        $ary = [1, 2, 1, 2, 3];
        $ary2 = [1, 2];
        $ary3 = [1, 2, 3];
        $ary4 = [1, 2, 3, 4];
        $idxOfAry2LastOccursInAry = findEnd($ary, $ary2);
        $this->assertEquals(2, $idxOfAry2LastOccursInAry);
        $this->assertEquals(2, findEnd($ary, $ary3));
        $this->assertNull(findEnd($ary, $ary4));
    }

    public function testFindFirstOf()
    {
        $ary2 = [10, 8, 2];
        $idxOfFirstAryElementInAry2 = findFirstOf($this->ary, $ary2);
        $this->assertEquals(1, $idxOfFirstAryElementInAry2);
        $ary3 = [10, 8];
        $this->assertNull(findFirstOf($this->ary, $ary3));
    }

    public function testSearch()
    {
        $ary2 = [2, 3, 4];
        $idxOfAry2InAry = search($this->ary, $ary2);
        $this->assertEquals(1, $idxOfAry2InAry);
        $aryNotConsistent = [2, 2, 4];
        $this->assertNull(search($this->ary, $aryNotConsistent));

        $this->assertEquals(0, search($this->ary, $this->ary));
        $aryTooLarge = range(1, 6);
        $this->assertNull(search($this->ary, $aryTooLarge));
    }

    public function testSearchN()
    {
        $ary = [1, 2, 2, 3];
        $this->assertNull(searchN($ary, 3, 2));
        $this->assertEquals(1, searchN($ary, 2, 2));
    }
}
?>
